File: app/src/main/java/name/monwf/customiuizer/mods/utils/WeatherDataController.java
package name.monwf.customiuizer.mods.utils;

import android.annotation.SuppressLint;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.database.Cursor;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;

import java.lang.ref.WeakReference;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class WeatherDataController {
    public static String weatherInfo = "";
    private static WeakReference<Context> weakReferenceContext;
    private static Runnable weakRefrenceRunnable;
    private static final ExecutorService sExecutor = Executors.newSingleThreadExecutor();
    private static final Handler sHandler = new Handler(Looper.getMainLooper());
    private static final Object sLock = new Object();
    private static BroadcastReceiver sTimeTickReceiver;
    private static Context sContext;

    private static void queryWeather() {
        Context ctx;
        synchronized (sLock) {
            ctx = weakReferenceContext != null ? weakReferenceContext.get() : null;
        }
        if (ctx == null) return;
        Cursor cursor = null;
        try {
            cursor = ctx.getContentResolver().query(
                Uri.parse("content://weather/actualWeatherData/1"), null, null, null, null);
            if (cursor != null && cursor.moveToFirst()) {
                String newWeather = "";
                int cursorColumnIndex = cursor.getColumnIndex("description");
                if (cursorColumnIndex >= 0) {
                    newWeather = cursor.getString(cursorColumnIndex);
                }
                cursorColumnIndex = cursor.getColumnIndex("temperature");
                if (cursorColumnIndex >= 0) {
                    newWeather += (" " + cursor.getString(cursorColumnIndex));
                }
                weatherInfo = newWeather;
            }
        } catch (Throwable ignore) {
        } finally {
            if (cursor != null) cursor.close();
        }
    }

    public static void refreshWeatherData(boolean forceRefresh) {
        sExecutor.execute(() -> {
            queryWeather();
            if (forceRefresh) {
                Runnable r;
                synchronized (sLock) {
                    r = weakRefrenceRunnable;
                    weakRefrenceRunnable = null;
                }
                if (r != null) {
                    sHandler.post(r);
                }
            }
        });
    }

    @SuppressLint("UnspecifiedRegisterReceiverFlag")
    public static void initContext(Context context, Runnable updateTimeRunnable) {
        synchronized (sLock) {
            weakReferenceContext = new WeakReference<>(context);
            weakRefrenceRunnable = updateTimeRunnable;
        }
        if (sTimeTickReceiver != null && sContext != null) {
            try {
                sContext.unregisterReceiver(sTimeTickReceiver);
            } catch (Throwable ignore) {}
        }
        sContext = context;
        sTimeTickReceiver = new BroadcastReceiver() {
            public void onReceive(Context context, Intent intent) {
                refreshWeatherData(false);
            }
        };
        sContext.registerReceiver(sTimeTickReceiver,
            new IntentFilter("android.intent.action.TIME_TICK"),
            Context.RECEIVER_NOT_EXPORTED);
        sHandler.postDelayed(() -> refreshWeatherData(true), 1800);
    }
}

File: app/src/main/java/tv/withaibuild/customiuizer/mods/utils/WeatherDataController.kt
package tv.withaibuild.customiuizer.mods.utils

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.database.Cursor
import android.net.Uri
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.lang.ref.WeakReference

object WeatherDataController {

    @JvmField
    var weatherInfo: String = ""

    private var weakReferenceContext: WeakReference<Context>? = null
    private var weakReferenceRunnable: Runnable? = null
    private var timeTickReceiver: BroadcastReceiver? = null
    private var context: Context? = null

    private val lock = Any()
    private var controllerScope = newScope()

    private fun newScope(): CoroutineScope =
        CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    private fun queryWeather() {
        val ctx: Context
        synchronized(lock) {
            ctx = weakReferenceContext?.get() ?: return
        }

        var cursor: Cursor? = null
        try {
            cursor = ctx.contentResolver.query(
                Uri.parse("content://weather/actualWeatherData/1"),
                null, null, null, null
            )
            if (cursor != null && cursor.moveToFirst()) {
                var newWeather = ""
                var columnIndex = cursor.getColumnIndex("description")
                if (columnIndex >= 0) {
                    newWeather = cursor.getString(columnIndex)
                }
                columnIndex = cursor.getColumnIndex("temperature")
                if (columnIndex >= 0) {
                    newWeather += (" " + cursor.getString(columnIndex))
                }
                weatherInfo = newWeather
            }
        } catch (ignored: Throwable) {
        } finally {
            cursor?.close()
        }
    }

    @JvmStatic
    fun refreshWeatherData(forceRefresh: Boolean) {
        controllerScope.launch {
            withContext(Dispatchers.IO) { queryWeather() }
            if (forceRefresh) {
                val r: Runnable?
                synchronized(lock) {
                    r = weakReferenceRunnable
                    weakReferenceRunnable = null
                }
                r?.run()
            }
        }
    }

    @SuppressLint("UnspecifiedRegisterReceiverFlag")
    @JvmStatic
    fun initContext(context: Context, updateTimeRunnable: Runnable) {
        synchronized(lock) {
            weakReferenceContext = WeakReference(context)
            weakReferenceRunnable = updateTimeRunnable
        }

        // Cancel any pending work from a previous context and start fresh.
        controllerScope.cancel()
        controllerScope = newScope()

        val oldContext: Context?
        val oldReceiver: BroadcastReceiver?
        synchronized(lock) {
            oldContext = this.context
            oldReceiver = timeTickReceiver
        }
        oldReceiver?.let {
            try {
                oldContext?.unregisterReceiver(it)
            } catch (ignored: Throwable) {
            }
        }

        this.context = context
        timeTickReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                refreshWeatherData(false)
            }
        }
        context.registerReceiver(
            timeTickReceiver,
            IntentFilter("android.intent.action.TIME_TICK"),
            Context.RECEIVER_NOT_EXPORTED
        )

        controllerScope.launch {
            delay(1800)
            refreshWeatherData(true)
        }
    }
}
